### ISSUE TITLE
Multi-region deployments

### ISSUE THREAD

**Issue by pmuens** [serverless/serverless#4123]
# This is a Feature Proposal

## Description

Add support for multi-region deployments (originally brought up by @nikgraf 👍).

**Comment by gianarb** [comment]
I think I have a use case then maybe can help. I am also looking to find a solution to avoid a too craft approach.

I have a lambda function triggered every time than a new EC2 change status to running or terminated. This function takes the ec2's public_ip and if the status is running it adds a new item in DynamoDB, if it's terminated it deletes the item.
The database is always the same. Let's say in London. Doesn't matter where the EC2 the Dynamo is always in London.

CloudWatch and Lambda are "for regions" it means that I need to find a way to deploy my function in the other regions.

Plus: I am using a CloudFormation stack configurated in `serverless.yml` it creates the DynamoDB table, Cloud Watch role and so on. The problem here is, let's say that I have a magic command or trick to deploy in multiple regions, but I also need to have different CloudFormation stack, because in my case I need to update/create DynamoDB table only in London.

Do you have some feedback?

**Comment by alexcasalboni** [comment]
@pmuens I think this will be even more useful given the announcement of DynamoDB Global Tables. I'm not sure how/when they'll be supported by CloudFormation, but I'd love to NOT have to manually deploy to N different regions.

**Comment by orfin** [comment]
@pmuens @alexcasalboni keeping in mind Lambda@Edge limitations (for us, the most blocking one: no way to run Lambdas in private VPC -- so no way to access internal Databases through inter-region VPC peering), we also recently started to think about such feature.

Are there any plans to have it included directly in Serverless Framework or should we better think about serverless plugin to handle that case? WDYT?

Some references:
* CloudFormation Stack Sets could be used: https://aws.amazon.com/blogs/aws/use-cloudformation-stacksets-to-provision-resources-across-multiple-aws-accounts-and-regions/
* Apex is already working on that: https://github.com/apex/up/issues/134

Thanks, Antoni

**Comment by s-chand** [comment]
Need to use AWS services that aren't available in the current region alongside services in another region

**Comment by localredhead** [comment]
Definitely need this

**Comment by dorsegal** [comment]
+1

**Comment by nibynool** [comment]
My use case is a Lambda@Edge function to do some URL rewriting so my Cloudfront distribution can use CloudFrontOriginIdentity.

I've got everything running in us-west-2, but assigning a LambdaFunctionAssociation in a CloudFrontDistribution requires the function to be located in us-east-1.

I'd love to be able to deploy a lambda function to a specific region while allowing my resources and other functionality to be deployed elsewhere.

**Comment by Imran99** [comment]
This would be great for us too!

We currently run our apis multi region in order to support our global user base and compliance / legal requirements around data locality. Deploying to each region in serial adds quite a bit of time to our deployments (we operate in 4 regions).

We started using circleci parallelisation to address this (deploy to each region using a separate worker / container in parallel) but we still have the slow to deploy multi region issue in our lower dev environment where we are doing more frequent deploys on our personal dev stacks. The circle ci config to do this is a little less straightforward than it could be too.

We also found that simply spawning multiple builds in parallel (e.g. `sls deploy --region us-east-1 & sls deploy --region eu-west-1`) didn't play nice with serverless due to the state and temp files in the `.serverless` directory being overwritten.

Something simple like:
```sh
sls deploy --region eu-west-1,us-east-1,xxx
```

which just deployed a Cloudformation stack to each region in parallel would be a huge help.

**Comment by Imran99** [comment]
> I'd love to be able to deploy a lambda function to a specific region while allowing my resources and other functionality to be deployed elsewhere.

@nibynool You can do this now by putting your regional resources in a seperate yml say `yml/us-east-1/functions.yml` and referencing it from your `serverless.yml`:

```yml
functions:
  ${file(./yml/${self:opt.region}/functions.yml)}
```

Then just deploy `sls deploy --region us-east-1`

**Comment by garethmcc** [comment]
Perhaps a cleaner way to do this would be to have the `region` property be an array so that you can just specify multiple regions. Would still be tricky I feel to get parallel deployments happening but at least that would simplify the process somewhat.

**Comment by iamtheworstdev** [comment]
This would be great for those of us who control most of our 'infrastructure' with terraform and then use SLS to deploy strictly lambdas+domain-manager .. I have multi region APIs and would like to just pass in a handful of regions and let SLS deploy the regions and attach to the appropriate custom domain.

**Comment by RNag** [comment]
I agree, I think a comma separated option to `-r` would be nice. We have a use case where we deploy the same code / infrastructure to multiple regions, and  currently we have to wait for one regional deployment to finish before the next starts. It would be nice to run the serverless deploy in parallel 🙂

**Comment by mrowles** [comment]
> My use case is a Lambda@Edge function to do some URL rewriting so my Cloudfront distribution can use CloudFrontOriginIdentity.
> 
> I've got everything running in us-west-2, but assigning a LambdaFunctionAssociation in a CloudFrontDistribution requires the function to be located in us-east-1.
> 
> I'd love to be able to deploy a lambda function to a specific region while allowing my resources and other functionality to be deployed elsewhere.

This our our use case too, glad to see it's not yet implemented as per comment above and there is a work around :) Keen to see a function-specific region setting

**Comment by tomhatzer** [comment]
> > My use case is a Lambda@Edge function to do some URL rewriting so my Cloudfront distribution can use CloudFrontOriginIdentity.
> > I've got everything running in us-west-2, but assigning a LambdaFunctionAssociation in a CloudFrontDistribution requires the function to be located in us-east-1.
> > I'd love to be able to deploy a lambda function to a specific region while allowing my resources and other functionality to be deployed elsewhere.
> 
> This our our use case too, glad to see it's not yet implemented as per comment above and there is a work around :) Keen to see a function-specific region setting

I'd have the same use-case. Lambda function in us-east-1 but the rest would be in eu-central-1.

**Comment by uncompiled** [comment]
This feature would be amazing since our current deployment process invokes `serverless deploy --region` multiple times.  Our resources are identical, but deployed in every region.

**Comment by dseeker** [comment]
having the same use case as @tomhatzer, main lambda in eu-central-1 being deploy through serverless pro. trying to include an edge lambda in the same project which can only happen on us-east-1
most logical would be to be able to specify region under each function.

**Comment by itaykat** [comment]
Any updates on that one?

**Comment by juvaly** [comment]
> My use case is a Lambda@Edge function to do some URL rewriting so my Cloudfront distribution can use CloudFrontOriginIdentity.
> 
> I've got everything running in us-west-2, but assigning a LambdaFunctionAssociation in a CloudFrontDistribution requires the function to be located in us-east-1.
> 
> I'd love to be able to deploy a lambda function to a specific region while allowing my resources and other functionality to be deployed elsewhere.

This was also my case. I solved it by moving my Lambda@Edge functions into a separate config file in the same project (serverless-edge.yml), and then I deploy twice:

```sls deploy --region eu-central-1 && yarn deploy --region us-east-1 --config serverless-edge.yml```

**Comment by medikoo** [comment]
This looks closely related to https://github.com/serverless/serverless/issues/4018

What's not clear is that wether this issue is about:
- Deploying same stack to multiple regions (so we deploy all stack resources to multi-regions). Effect of that would be same as deploying twice same stack but to two different regions (this somewhat could be automated I believe)

or 

- Deciding per resource to which region they're being deployed. For this case, it'll be good to confirm weather that's possible to achieve with regular CloudFormation stack deployment (as stack on its own needs to go to specific region I assume that such multi-region deployments can only be achieved via StackSets -> https://github.com/serverless/serverless/issues/4018)

**Comment by darylteo** [comment]
Sorry to resurrect an old issue but:

@medikoo I believe this request is to solve the 2nd problem. The use case encountered by myself (and similar to others) is when deploying for Lambda@Edge, which specifically requires the Edge Lambda to be deployed in us-east-1, while other existing infrastructure exist in other regions (specifically VPCs). I would argue that this should be considered part of the "same" stack. 

No experience with StackSets to comment on the other issue and it's relatedness. 

I initially thought I could override any properties on the top level on the function level but obviously that's not the case and I got configuration warnings. 

The workaround listed works ok but does require 2 calls, which in itself isn't a massive deal. So I can understand if it's a low priority. Happy to have found a workaround at least.

**Comment by medikoo** [comment]
@darylteo Currently Framework doesn't provide any means to deploy multiple stacks, but we're discussing introducing a multi-provider engine, which will introduce a bigger shift and will allow such deployments. 

There's a loose technical proposal (nothing set in stone) https://github.com/serverless/serverless/issues/9583 please follow that. If there'll be any progress on this it'll definitely be indicated in that issue.
